augur/curate/apply_date_bounds.py

```diff
@@ -0,0 +1,231 @@
+                print_err(f"WARNING: {error}")
+                failures.append(error)
+                continue
+            if args.failure_reporting is DataErrorMethod.ERROR_ALL:
+                failures.append(error)
+                continue
+            else:
+                raise ValueError(f"Encountered unhandled failure reporting method: {args.failure_reporting!r}")
+        yield record
+
+    if args.failure_reporting is not DataErrorMethod.SILENT and failures:
+        if args.failure_reporting is DataErrorMethod.ERROR_ALL:
+            raise AugurError(dedent(f"""\
+                Unable to apply bounds. All errors:
+                {indented_list(map(str, failures), "                ")}"""))
+
+
+def validate_arguments(args: argparse.Namespace):
+    if not args.lower_bound and not args.upper_bound:
+        raise AugurError("At least one of --lower-bound and --upper-bound is required.")
+
+
+class Record:
+    """
+    Helper class to wrap a record, its id, and arguments for ease of error handling and small functions.
+    """
+
+    def __init__(self, data: RecordType, record_id: int) -> None:
+        self.data = data
+        self.id = record_id
+
+    def get_bounded_date(self, args: argparse.Namespace) -> str:
+        """
+        Returns a date string representing the date with bounds applied. bounded interval for the given record.
+        """
+        start, end = self.convert_date_to_range(args.date_field)
+
+        lower_bound, upper_bound = self.get_bounds(args.lower_bound, args.upper_bound)
+
+        # If any ends are unbounded, return the original date
+        if (start == float("-inf") and lower_bound is None) or \
+           (end   == float( "inf") and upper_bound is None):
+            return self.data[args.date_field]
+
+        # Error if start or end are out of bounds.
+        if lower_bound and start < lower_bound and end < lower_bound:
+            self.raise_data_error(
+                f"{args.date_field!r}={self.data[args.date_field]!r} "
+                f"is earlier than the lower bound of "
+                f"{args.lower_bound!r}={self.data[args.lower_bound]!r}"
+            )
+        if upper_bound and start > upper_bound and end > upper_bound:
+            self.raise_data_error(
+                f"{args.date_field!r}={self.data[args.date_field]!r} "
+                f"is later than the upper bound of "
+                f"{args.upper_bound!r}={self.data[args.upper_bound]!r}"
+            )
+
+        # If the target date overlaps with the bounds, apply the bounds.
+        # The start should be no earlier than the lower bound
+        # and the end should be no later than the upper bound.
+        if lower_bound:
+            start = max(start, lower_bound)
+        if upper_bound:
+            end = min(end, upper_bound)
+
+        # ISO 8601 interval in <start>/<end> format
+        return f"{datestring_from_numeric(start)}/{datestring_from_numeric(end)}"
```

Comment:
I don't think ISO 8601 explicitly supports or prohibits negative dates, but should they be allowed here?

```
$ echo '{"date": "-200.0"}' | ./bin/augur curate apply-date-bounds --date-field date --lower-bound -300.0 
--upper-bound 2025-01-01
{"date": "-200-01-01/-200-01-01"}
```